docs/man/mda-lmtp.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "mda-lmtp 1"
.TH mda-lmtp 1 2018-04-02 "" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
mda\-lmtp \- MDA that uses LMTP to do the mail delivery
.SH SYNOPSIS
.IX Header "SYNOPSIS"
mda\-lmtp
[\fB\-addr_network\fR \fInet\fR]
\&\fB\-addr\fR \fIaddr\fR
\&\fB\-d\fR \fIrecipient\fR
\&\fB\-f\fR \fIfrom\fR
.SH DESCRIPTION
.IX Header "DESCRIPTION"
mda\-lmtp is a very basic MDA that uses LMTP to do the mail delivery.
.PP
It takes command line arguments similar to maildrop or procmail, reads an
email via standard input, and sends it over the given LMTP server.  Supports
connecting to LMTP servers over UNIX sockets and TCP.
.PP
It can be used when your mail server does not support LMTP directly.
.SH EXAMPLE
.IX Header "EXAMPLE"
\&\fBmda\-lmtp\fR \fI\-\-addr=localhost:1234\fR \fI\-f juan@casa\fR \fI\-d jose\fR < email
.SH OPTIONS
.IX Header "OPTIONS"
.IP "\fB\-addr\fR \fIaddress\fR" 8
.IX Item "-addr address"
LMTP server address to connect to.
.IP "\fB\-addr_network\fR \fInetwork\fR" 8
.IX Item "-addr_network network"
Network of the LMTP address (e.g. \fIunix\fR or \fItcp\fR). If not present, it will
be autodetected from the address itself.
.IP "\fB\-d\fR \fIrecipient\fR" 8
.IX Item "-d recipient"
Recipient.
.IP "\fB\-f\fR \fIfrom\fR" 8
.IX Item "-f from"
Whom the message is from.
.SH "RETURN VALUES"
.IX Header "RETURN VALUES"
.IP \fB0\fR 8
.IX Item "0"
success
.IP \fB75\fR 8
.IX Item "75"
temporary failure
.IP \fIother\fR 8
.IX Item "other"
permanent failures (usually indicate misconfiguration)
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBchasquid\fR\|(1)


cmd/mda-lmtp/mda-lmtp.go
// mda-lmtp is a very basic MDA that uses LMTP to do the delivery.
//
// See the usage below for details.
package main

import (
	"flag"
	"fmt"
	"io"
	"net"
	"net/textproto"
	"os"
	"strings"

	"golang.org/x/net/idna"
)

// Command-line flags
var (
	fromwhom  = flag.String("f", "", "Whom the message is from")
	recipient = flag.String("d", "", "Recipient")

	addrNetwork = flag.String("addr_network", "",
		"Network of the LMTP address (e.g. unix or tcp)")
	addr = flag.String("addr", "", "LMTP server address")

	toPuny = flag.Bool("to_puny", false,
		"Encode addresses using punycode")
)

func usage() {
	fmt.Fprintf(os.Stderr, `
mda-lmtp is a very basic MDA that uses LMTP to do the mail delivery.

It takes command line arguments similar to maildrop or procmail, reads an
email via standard input, and sends it over the given LMTP server.
Supports connecting to LMTP servers over UNIX sockets and TCP.

It can be used when your mail server does not support LMTP directly.

Example of use:
$ mda-lmtp --addr localhost:1234 -f juan@casa -d jose < email

Flags:
`)
	flag.PrintDefaults()
}

// Exit with EX_TEMPFAIL.
func tempExit(format string, args ...any) {
	fmt.Printf(format+"\n", args...)
	// 75 = EX_TEMPFAIL "temporary failure" exit code (sysexits.h).
	os.Exit(75)
}

func permExit(format string, args ...any) {
	fmt.Printf(format+"\n", args...)
	os.Exit(2)
}

func main() {
	var err error
	flag.Usage = usage
	flag.Parse()

	if *addr == "" {
		permExit("No LMTP server address given (use --addr)")
	}

	if *toPuny {
		*fromwhom, err = idna.ToASCII(*fromwhom)
		if err != nil {
			permExit("cannot puny-encode from: %v", err)
		}
		*recipient, err = idna.ToASCII(*recipient)
		if err != nil {
			permExit("cannot puny-encode recipient: %v", err)
		}
	}

	// Try to autodetect the network if it's missing.
	if *addrNetwork == "" {
		*addrNetwork = "tcp"
		if strings.HasPrefix(*addr, "/") {
			*addrNetwork = "unix"
		}
	}

	conn, err := net.Dial(*addrNetwork, *addr)
	if err != nil {
		tempExit("Error connecting to (%s, %s): %v",
			*addrNetwork, *addr, err)
	}

	tc := textproto.NewConn(conn)

	// Expect the hello from the server.
	_, _, err = tc.ReadResponse(220)
	if err != nil {
		tempExit("Server greeting error: %v", err)
	}

	hostname, err := os.Hostname()
	if err != nil {
		tempExit("Could not get hostname: %v", err)
	}

	if *fromwhom == "<>" {
		*fromwhom = ""
	}
	if *recipient == "<>" {
		*recipient = ""
	}

	cmd(tc, 250, "LHLO %s", hostname)
	cmd(tc, 250, "MAIL FROM:<%s>", *fromwhom)
	cmd(tc, 250, "RCPT TO:<%s>", *recipient)
	cmd(tc, 354, "DATA")

	w := tc.DotWriter()
	_, err = io.Copy(w, os.Stdin)
	w.Close()
	if err != nil {
		tempExit("Error writing DATA: %v", err)
	}

	// This differs from SMTP: here we get one reply per recipient, with the
	// result of the delivery. Since we deliver to only one recipient, read
	// one code.
	_, _, err = tc.ReadResponse(250)
	if err != nil {
		tempExit("Delivery failed remotely: %v", err)
	}

	cmd(tc, 221, "QUIT")

	tc.Close()
}

// cmd sends a command and checks it matched the expected code.
func cmd(conn *textproto.Conn, expectCode int, format string, args ...any) {
	id, err := conn.Cmd(format, args...)
	if err != nil {
		tempExit("Sent %q, got %v", fmt.Sprintf(format, args...), err)
	}
	conn.StartResponse(id)
	defer conn.EndResponse(id)

	_, _, err = conn.ReadResponse(expectCode)
	if err != nil {
		tempExit("Sent %q, got %v", fmt.Sprintf(format, args...), err)
	}
}
